File: SQLi Example 1/SQLi1_safe.php
<?php

/*
    This code contains an SQL query string that is protected against SQL injection
    attacks. It is not meant to be flagged by a SAST tool for possible SQL injection.

    Instead of concatenating user input directly into the SQL query string,
    we use prepared statements with placeholders.
*/

// Database connection details
$host = "localhost";
$dbname = "exampledb";
$username = "root";
$password = "password";

try {
    // Create a new PDO connection
    $conn = new PDO("mysql:host=$host;dbname=$dbname", $username, $password);
    $conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

    // Prompt user for input (via CLI for simplicity)
    echo "Enter username: ";
    $user = trim(fgets(STDIN)); // Read username from input

    echo "Enter password: ";
    $pass = trim(fgets(STDIN)); // Read password from input

    // Safe query construction
    $query = "SELECT * FROM users WHERE username = :username AND password = :password";

    // Execute the query
    $stmt = $conn->prepare($query);
    $stmt->bindParam(':username', $user, PDO::PARAM_STR);
    $stmt->bindParam(':password', $pass, PDO::PARAM_STR);
    $stmt->execute();
    
    $results = $stmt->fetchAll();
    
    if (count($results) > 0) {
        echo "Login successful!" . PHP_EOL;
    } else {
        echo "Invalid username or password." . PHP_EOL;
    }

} catch (PDOException $e) {
    echo "Database error: " . $e->getMessage() . PHP_EOL;
}
?>

File: SQLi Example 1/SQLi1_safe.java
/*
    This code contains an SQL query string that is protected against SQL injection
    attacks. It is not meant to be flagged by a SAST tool for possible SQL injection.

    Instead of concatenating user input directly into the SQL query string,
    we use prepared statements with parameterized queries.
*/

import java.sql.*;
import java.util.Scanner;

public class SQLi1_safe {
    public static void main(String[] args) {
        // Set up a Scanner to read user input
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter username: ");
        String username = scanner.nextLine(); // Read username from user input

        System.out.println("Enter password: ");
        String password = scanner.nextLine(); // Read password from user input

        try {
            // Establish a database connection
            Connection connection = DriverManager.getConnection("jdbc:mysql://localhost:3306/mydatabase", "root", "password");

            // Safe query construction
            String query = "SELECT * FROM users WHERE username = ? AND password = ?";

            // Create a statement and execute it
            PreparedStatement preparedStatement = connection.prepareStatement(query);
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            ResultSet resultSet = preparedStatement.executeQuery();

            if (resultSet.next()) {
                System.out.println("Login successful!");
            } else {
                System.out.println("Invalid username or password.");
            }

            // Close resources
            resultSet.close();
            preparedStatement.close();
            connection.close();
            scanner.close();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
